src/esi/Expression.cc

```diff
@@ -280,6 +284,7 @@ evalor(stackmember * stack, int *depth, int whereAmI, stackmember * candidate)
 
     srv.precedence = 1;
 
+    Must2(*depth < STACK_DEPTH_LIMIT-1, "expression too complex");
```

Comment:
The code below this (and similar) Must2() does no need `*depth` to be smaller than `STACK_DEPTH_LIMIT-1`. Being smaller than `STACK_DEPTH_LIMIT` is enough AFAICT.
```suggestion
    Must2(*depth < STACK_DEPTH_LIMIT, "ESI expression too complex");
```

Reply:
The expression code relies on the stack[depth] member being an initialized stackmember object in some places. While it looks like STACK_DEPTH_LIMIT is enough, some test cases hit a segfault in later evaluation code. So I am being a bit cautious for these initial new exceptions.

Reply:
I do not recommend adding exceptions that do not match the seemingly associated code because it makes the reader question both the exception and the code. If some other code needs a different (e.g., more strict) exception, add two exceptions. If we should check a seemingly unrelated condition for some obscure reason, add a comment explaining it. Hopefully, we can do the former and, hence, do not need to do the latter.

Reply:
IMO this is resolved with the addition of stackpush() function requested earlier. That function is clearly documented as ensuring depth value is valid before *and after* the push.

Reply:
Yes, this is now resolved, at least partially. If I find specific places in the new code that need further work, I will mark them as such.